cbmc/proofs/proof_guide.md

```diff
@@ -0,0 +1,399 @@
+
+This produces `logs/result.txt` in plaintext format.
+
+Before pushing a new proof for a new function, make sure that _all_ proofs run OK from the `cbmc/proofs` directory with
+
+```
+export MLKEM_K=3 # or 2 or 4...
+./run-cbmc-proofs.py --summarize -j8
+```
+
+Change the "8" to match the number of CPU cores that you have available.
+
+## Worked Example - proving poly_tobytes()
+
+This section follows the recipe above, and adds actual settings, contracts and command to prove the `poly_tobytes()` function.
+
+### Populate a proof directory
+
+The proof directory is `cbmc/proofs/poly_tobytes`
+
+### Update Makefile
+
+The significant changes are:
+```
+HARNESS_FILE = poly_tobytes_harness
+PROOF_UID = poly_tobytes
+PROJECT_SOURCES += $(SRCDIR)/mlkem/poly.c
+CHECK_FUNCTION_CONTRACTS=$(MLKEM_NAMESPACE)poly_tobytes
+USE_FUNCTION_CONTRACTS=
+FUNCTION_NAME = $(MLKEM_NAMESPACE)poly_tobytes
+```
+Note that `USE_FUNCTION_CONTRACTS` is left empty since `poly_tobytes()` is a leaf function that does not call any other functions at all.
+
+### Update harness function
+
+`poly_tobytes()` has a simple API, requiring two parameters, so the harness function is:
+
+```
+void harness(void) {
+  poly a;
+  uint8_t r[MLKEM_POLYBYTES];
+
+  /* Contracts for this function are in poly.h */
+  poly_tobytes(r, &a);
+}
+```
+
+### Top-level contracts
+
+The comments on `poly_tobytes()` give us a clear hint:
+
+```
+ * Arguments:   INPUT:
+ *              - a: const pointer to input polynomial,
+ *                with each coefficient in the range [0,1,..,Q-1]
+ *              OUTPUT
+ *              - r: pointer to output byte array
+ *                   (of MLKEM_POLYBYTES bytes)
+```
+So we need to write a REQUIRES contract to constrain the ranges of the coefficients denoted by the parameter `a`. There is no constraint on the output byte array, other than it must be the right length, which is given by the function prototype.
+
+We can use the macros in `cbmc.h` to help, thus:
+
+```
+void poly_tobytes(uint8_t r[MLKEM_POLYBYTES], const poly *a)
+     // clang-format off
+REQUIRES(a != NULL && IS_FRESH(a, sizeof(poly)))
+REQUIRES(ARRAY_IN_BOUNDS(int, k, 0, (MLKEM_N - 1), a->coeffs, 0, (MLKEM_Q - 1)))
+ASSIGNS(OBJECT_WHOLE(r));
+// clang-format on
+```
+
+### Interior contracts and loop invariants
+
+`poly_tobytes` has a reasonable simple, single loop statement:
+
+```
+  for (unsigned int i = 0; i < MLKEM_N / 2; i++)
+  { ... }
+```
+
+A candidate loop contract needs to state that:
+1. The loop body assigns to variable `i` and the whole object pointed to by `r`.
+2. Loop counter variable `i` is in range 0 .. MLKEM_N / 2 at the point of the loop invariant (remember the pattern above).
+3. The loop terminates because the expression `MLKEM_N / 2 - i` decreases on every iteration.
+
+Therefore, we add:
+
+```
+  for (unsigned int i = 0; i < MLKEM_N / 2; i++)
+      // clang-format off
+  ASSIGNS(i, OBJECT_WHOLE(r))
+  INVARIANT(i >= 0)
```

Comment:
Isn't this sort of pointless as it's unsigned? Should we change the loop counter to `int` in those cases?

Reply:
Yes... it's pointless for an unsigned type, but just being explicit.

Reply:
Added a note that it can be left out.